File: go/expr/expr.go
package expr

// EvaluateExpr should return 64-bit integer result of evaluating a *valid* expression in expr.
// Expression can contain 1-digit integers '0'..'9', brackets '(', ')' and
// binary operators '+', '-', '*', '/'. Multiply and division have higher priority than add/sub.
// Division by zero produces zero result. For example, expression '1+2*(3/0-4)' evaluates to -7.
func EvaluateExpr(expr string) int {
	return rpnEval(rpnFromExpr(expr))
}

func rpnFromExpr(expr string) []byte {
	var rpn, stack []byte

	pri := map[byte]int{
		'(': 0,
		')': 1,
		'+': 2,
		'-': 2,
		'*': 3,
		'/': 3,
	}

	for _, op := range []byte(expr) {
		switch {
		case op >= '0' && op <= '9':
			rpn = append(rpn, op)

		case op == '(':
			stack = append(stack, op)

		default:
			sp := len(stack) - 1
			for ; sp >= 0 && pri[stack[sp]] >= pri[op]; sp-- {
				rpn = append(rpn, stack[sp])
			}

			if op == ')' {
				stack = stack[:sp]
			} else {
				stack = append(stack[:sp+1], op)
			}
		}
	}

	for sp := len(stack) - 1; sp >= 0; sp-- {
		rpn = append(rpn, stack[sp])
	}

	return rpn
}

func rpnEval(rpn []byte) int {
	var stack []int

	for _, op := range rpn {
		if op >= '0' && op <= '9' {
			stack = append(stack, int(op-'0'))
			continue
		}

		sp := len(stack) - 2
		a, b, c := stack[sp], stack[sp+1], 0

		switch op {
		case '+':
			c = a + b
		case '-':
			c = a - b
		case '*':
			c = a * b
		case '/':
			if b != 0 {
				c = a / b
			}
		}

		stack = append(stack[:sp], c)
	}

	return stack[0]
}

File: java/src/expr/Expr.java
package expr;

import java.util.*;

public class Expr {
	// evaluateExpr should return 64-bit integer result of evaluating expression in expr.
	// Expression can contain 1-digit integers '0'..'9', brackets '(', ')' and
	// binary operators '+', '-', '*', '/'. Multiply and division have higher priority than add/sub.
	// Division by zero produces zero result. For example, expression '1+2*(3/0-4)' evaluates to -7.
	public static long evaluateExpr(String expr) {
		char [] pol = new char[expr.length()];
		char [] temp = new char[expr.length()];
		int r = 0, t = 0;
		long [] stack = new long[expr.length()];

		for (char c : expr.toCharArray()) {
			if (Character.isDigit(c)) {
				pol[r++] = c;
				continue;
			}

			while (c != '(' && t > 0 && priority(c) <= priority(temp[t - 1])) {
				pol[r++] = temp[--t];
			}

			if (c == ')') {
				t--;
			} else {
				temp[t++] = c;
			}
		}

		while (t > 0) {
			pol[r++] = temp[--t];
		}

		t = 0;
		for (int i = 0; i < r; i++) {
			char c = pol[i];
			if (Character.isDigit(c)) {
				stack[t++] = c - '0';
				continue;
			}

			long b = stack[--t];
			long a = stack[--t];
			switch (c) {
				case '*':
					stack[t++] = a * b;
					break;
				case '/':
					if (b == 0) {
						stack[t++] = 0;
					} else {
						stack[t++] = a/b;
					}
					break;
				case '+':
					stack[t++] = a + b;
					break;
				case '-':
					stack[t++] = a - b;
					break;
			}
		}

		return stack[0];
	}

	private static int priority(char c) {
		switch (c) {
			case '*':
			case '/':
				return 3;
			case '+':
			case '-':
				return 2;
			case ')':
				return 1;
			case '(':
				return 0;
		}
		return 0;
	}

}
